File: src/main/java/pl/allegro/experiments/chi/chiserver/infrastructure/InMemoryExperimentsRepository.java
package pl.allegro.experiments.chi.chiserver.infrastructure;

import com.google.common.collect.ImmutableList;
import pl.allegro.experiments.chi.chiserver.domain.experiments.Experiment;
import pl.allegro.experiments.chi.chiserver.domain.experiments.ExperimentDefinition;
import pl.allegro.experiments.chi.chiserver.domain.experiments.ExperimentsRepository;

import java.util.*;
import java.util.concurrent.ConcurrentHashMap;
import java.util.stream.Collectors;

public class InMemoryExperimentsRepository implements ExperimentsRepository {
    private final Map<String, Experiment> experiments = new ConcurrentHashMap<>();

    public InMemoryExperimentsRepository(Collection<Experiment> experiments) {
        experiments.forEach(e -> this.experiments.put(e.getId(),e));
    }

    @Override
    public Optional<Experiment> getExperiment(String experimentId) {
        return Optional.ofNullable(experiments.get(experimentId));
    }

    public void save(Experiment experiment) {
        experiments.put(experiment.getId(), experiment);
    }

    @Override
    public void save(ExperimentDefinition experimentDefinition) {
        experiments.put(experimentDefinition.getId(), experimentDefinition.toExperiment());
    }

    @Override
    public void delete(String experimentId) {
        remove(experimentId);
    }

    @Override
    public List<Experiment> getAll() {
        return ImmutableList.copyOf(experiments.values());
    }

    @Override
    public List<Experiment> assignable() {
        return ImmutableList.copyOf(experiments.values().stream()
                .filter(Experiment::isAssignable)
                .collect(Collectors.toList()));
    }

    public Set<String> experimentIds() {
        return experiments.keySet();
    }

    public void remove(String experimentId) {
        experiments.remove(experimentId);
    }
}

File: src/main/kotlin/pl/allegro/experiments/chi/chiserver/infrastructure/InMemoryExperimentsRepository.kt
package pl.allegro.experiments.chi.chiserver.infrastructure

import com.google.common.collect.ImmutableList
import pl.allegro.experiments.chi.chiserver.domain.experiments.Experiment
import pl.allegro.experiments.chi.chiserver.domain.experiments.ExperimentsRepository

open class InMemoryExperimentsRepository(experiments: Collection<Experiment>) : ExperimentsRepository {
    private val experiments: MutableMap<String, Experiment> = experiments
            .associateBy { it.id }
            .toMutableMap()

    override fun getExperiment(id: String): Experiment? {
        return experiments.get(id)
    }

    override fun save(experiment: Experiment) {
        experiments.put(experiment.id, experiment)
    }

    internal fun remove(experimentId: String) {
        experiments.remove(experimentId)
    }

    override fun delete(experimentId: String) {
        remove(experimentId)
    }

    internal fun experimentIds(): Set<String> = experiments.keys

    override fun getAll() : List<Experiment> = ImmutableList.copyOf(experiments.values)

    override fun assignable(): List<Experiment> {
        return ImmutableList.copyOf(experiments.values.filter { it.isAssignable() })
    }
}
